Tests/NIOTests/EventLoopFutureTest.swift

```diff
@@ -82,6 +82,59 @@ class EventLoopFutureTest : XCTestCase {
             XCTFail("error of wrong type \(e)")
         }
     }
+    
+    func testReduceWithAllSuccesses() throws {
+        let eventLoop = EmbeddedEventLoop()
+        let promises: [EventLoopPromise<Int>] = (0..<5).map { (_: Int) in eventLoop.newPromise() }
+        let futures = promises.map { $0.futureResult }
+        
+        let fN: EventLoopFuture<[Int]> = EventLoopFuture<[Int]>.reduce([], futures, eventLoop: eventLoop, reducer: {$0 + [$1]})
+        for i in 1...5 {
+            promises[i - 1].succeed(result: (i))
+        }
+        let results = try fN.wait()
+        XCTAssertEqual(results, [1, 2, 3, 4, 5])
+    }
+    
+    func testReduceWithAllFailures() throws {
+        struct E: Error {}
+        let eventLoop = EmbeddedEventLoop()
+        let promises: [EventLoopPromise<Int>] = (0..<100).map { (_: Int) in eventLoop.newPromise() }
+        let futures = promises.map { $0.futureResult }
+        
+        let fN: EventLoopFuture<Int> = EventLoopFuture<Int>.reduce(0, futures, eventLoop: eventLoop, reducer: +)
+        _ = promises.map { $0.fail(error: E()) }
+        do {
+            _ = try fN.wait()
+            XCTFail("should've thrown an error")
+        } catch _ as E {
+            /* good */
+        } catch let e {
+            XCTFail("error of wrong type \(e)")
+        }
+    }
+    
+    func testReduceWithOneFailure() throws {
+        struct E: Error {}
+        let eventLoop = EmbeddedEventLoop()
+        var promises: [EventLoopPromise<Int>] = (0..<100).map { (_: Int) in eventLoop.newPromise() }
+        _ = promises.map { $0.succeed(result: (1)) }
+        let failedPromise: EventLoopPromise<Int> = eventLoop.newPromise()
+        failedPromise.fail(error: E())
+        promises.append(failedPromise)
+        
+        let futures = promises.map { $0.futureResult }
+        
+        let fN: EventLoopFuture<Int> = EventLoopFuture<Int>.reduce(0, futures, eventLoop: eventLoop, reducer: +)
+        do {
+            _ = try fN.wait()
+            XCTFail("should've thrown an error")
+        } catch _ as E {
+            /* good */
+        } catch let e {
+            XCTFail("error of wrong type \(e)")
+        }
+    }
```

Comment:
Can we have one extra test that validates that the failure is fast? It should be easy enough: fail the first promise and confirm that the result future has completed immediately.